curp/src/sync_manager.rs

```diff
@@ -415,4 +455,120 @@ impl<C: Command + 'static> SyncManager<C> {
         }
         info!("background apply stopped");
     }
+
+    /// How long a candidate should wait before it starts another round of election
+    const CANDIDATE_TIMEOUT: Duration = Duration::from_secs(1);
+    /// How long a follower should wait before it starts a round of election (in millis)
+    const FOLLOWER_TIMEOUT: Range<u64> = 300..500;
+
+    /// Background election
+    async fn bg_election(
+        connects: Vec<Arc<Connect>>,
+        state: Arc<RwLock<State<C>>>,
+        last_rpc_time: Arc<RwLock<Instant>>,
+        heartbeat_trigger: Arc<Event>,
+    ) {
+        let role_trigger = state.read().role_trigger();
+        loop {
+            // only follower or candidate should run this task
+            while state.read().is_leader() {
+                role_trigger.listen().await;
+            }
+
+            let timeout = {
+                let mut rng = thread_rng();
+                match state.read().role() {
+                    ServerRole::Follower => {
+                        Duration::from_millis(rng.gen_range(Self::FOLLOWER_TIMEOUT))
+                    }
+                    ServerRole::Candidate => Self::CANDIDATE_TIMEOUT,
+                    ServerRole::Leader => continue, // leader should not vote
+                }
+            };
+            let next_check = last_rpc_time.read().to_owned() + timeout;
+            tokio::time::sleep_until(next_check).await;
+            if Instant::now() - *last_rpc_time.read() < timeout {
+                continue;
+            }
```

Comment:
It still requires periodically waking up to check, I guess a tokio::select helps here. Because we don't need to start a vote if heartbeat or append_entries arrives on time. We can ignore a state read lock here if you take my way.